### DISCUSSION TITLE
Improve startuptime significantly: Schedule vim.opt.clipboard

### DISCUSSION THREAD

**Discussion by abeldekat** [LazyVim/LazyVim#4112] [Recipes]
I recently posted the following on reddit( "Saving 40% startuptime for free" ):

> On arch linux, using xsel and nvim 0.10, clipboard.vim uses 43.17 on a total of 84.8...
> Would the clipboard still work as expected when scheduled?
>
> vim.schedule(function() vim.opt.clipboard = vim.env.SSH_TTY and "" or "unnamedplus" end)
>
> Indeed it does, and [startuptime](https://github.com/abeldekat/nvim_pde) is now 46.9

You can try this in your LazyVim config by adding the following to `lua/config/options.lua`:

```lua
vim.opt.clipboard = "" -- unset LazyVim's clipboard setting
vim.schedule(function() -- and schedule...
  vim.opt.clipboard = vim.env.SSH_TTY and "" or "unnamedplus"
end)
```

To see the difference, enable extra `util.startuptime`. Start lazy, and type `:StartupTime`.
You will not see the difference in the time reported by `lazy.nvim` on the dashboard.

**Comment by dpetka2001** [comment]
Great, I tried this and indeed there's a difference in the startup time measured by `vim-startuptime`. Bare in mind my laptop is really old (Dell model 2013, Intel Core i5-3337U, 6GB RAM, 500GB HDD 5400rpm).

Before this modification `vim-startuptime` measured 200-202ms total startup time, of which 142 were due to `clipboard.vim`. After the modification total startup time fell to almost 60 give or take and is more in line with what `lazy.nvim` also measures (50ms in my case although I believe there are some differences in how `lazy.nvim` measures startup time compared to `vim-startuptime`, but I don't really understand into depth what's going on). 

Also after this modification when opening Neovim without a file argument the Dashboard (on my old hardware) shows instantly, whereas before it took a very little time (0.1-0.2s) to redraw.

**Comment by folke** [comment]
Cool, I'll have a look.

Btw, in case you are using `bob`, then that is slow as hell as well.

To make it instant, you can use the following `fish` script. Put it in `~/.config/fish/functions/bob.fish`.

```fish
function bob --wraps bob
    command bob $argv
    set used (cat ~/.local/share/bob/used)
    set src ~/.local/share/bob/$used
    set dest ~/.local/share/bob-nvim
    test -L $dest
    and unlink $dest
    ln -s $src/nvim-linux64 $dest
end
```

And then add `~/.local/share/bob-nvim` to your path instead of what bob recommends.

Bob used to work exactly like this, but at some point it was changed to a wrapper executable that for some weird reason takes about `80ms` on my machine and I have a very fast laptop.

**Reply by dpetka2001** [reply]
Cool, I'll try to incorporate this in my fish shell as well since I use `bob/fish`. Tnx for sharing this :smile:

**Reply by folke** [reply]
Let me know if you can see the difference. It was very noticeable for me.
Make sure to open a new shell and run `bob use nightly` once.
Then verify `which nvim` to see that your path is correctly setup.

```
❯ which nvim
/home/folke/.local/share/bob-nvim/bin/nvim
```

**Reply by dpetka2001** [reply]
Yes, I can verify. Tnx a lot again :smiley:

**Comment by folke** [comment]
Also, to note. What `lazy.nvim` measures is cpu time by the Neovim process, so that would indeed not include any time spent in spawned processes.

**Comment by folke** [comment]
Just tested that clipboard thing and it doesn't make a difference for me on wayland:

```
❯ hyperfine "nvim +q"
Benchmark 1: nvim +q
  Time (mean ± σ):      39.9 ms ±   2.8 ms    [User: 28.1 ms, System: 8.0 ms]
  Range (min … max):    34.1 ms …  47.5 ms    64 runs
```

Same result when using `:StartupTime`. At least on wayland there seems to be no difference at all. Weird.

```
clipboard.vim          0.20    0.58 ▎
```

**Reply by dpetka2001** [reply]
Did you unset `vim.o.clipboard`? If you unset it in `options.lua` and then schedule it, then it shouldn't show on `Startuptime` at all.

**Reply by folke** [reply]
If I comment that line (without even the schedule), timing is the same.
This is why:
![image](https://github.com/user-attachments/assets/3d9ff2c7-b932-4233-af43-5f511139125e)

For `wl-copy` it doesn't execute it, but it does for `xsel`.

**Reply by folke** [reply]
This part
![image](https://github.com/user-attachments/assets/42ead009-d835-4544-a507-af5ad91746b0)

**Reply by dpetka2001** [reply]
You are correct. I uninstalled `xsel` and installed `xclip` and the difference is minimal if any at all. Without modification `clipboard.vim` takes only 0.49ms on my old laptop. So, I guess it has to do with the clipboard provider you're using and `xsel` is the culprit for this huge difference in startup time.

**Reply by abeldekat** [reply]
Amazing, for two reasons:

1. I had a look at `clipboard.vim` and completely missed that `s:cmd_ok`...;)
2. Why is `xsel` preferred over `xclip` when start-up performance can be affected that much?

The gain @dpetka reported is enormous in my opinion. 
Also, in my post on reddit, a user reported:

> Removed the option. Shaved 300ms from my startup time. Now i gotta get used to the new bindings. 

He uses `macos` and `pbcopy`. 300 ms!
First guess: For `macos` an internal `cache_enabled` is set to `0`

This is the commit in Neovim:

[clipboard: Prefer xsel (#20862)](https://github.com/neovim/neovim/pull/20918)

**Reply by abeldekat** [reply]
@folke,

Would a PR be useful? A considerable amount of `LazyVim` users probably use `xsel` or `pbcopy`. If so: It would be a tough one in my opinion. Its difficult to make sure that the `clipboard` settings of the (expert) user are not overwritten.

**Reply by dusty-phillips** [reply]
Some additional confirmation: On MacOS with pbcopy, my startup time changes from 101ms to 65ms. Hardware is a 2020 (pre-apple sillicon) imac pro.

**Reply by abeldekat** [reply]
@dusty-phillips Thanks!

I applied a fix in a local fork of LazyVim which I will post here asap...

**Reply by abeldekat** [reply]
I think [this](https://github.com/LazyVim/LazyVim/compare/main...abeldekat:LazyVim:main) solves the problem in a generic way. Its surprisingly simple. Just the solution I proposed here, but applied in `lua/lazyvim/config/init.lua`

```lua
  M.load("options")
  local clipboard = vim.opt.clipboard
  vim.opt.clipboard = "" -- PERF: reset
  vim.schedule(function()
    -- load later to avoid bad performance for "xsel" and "pbcopy"
    vim.opt.clipboard = clipboard
  end)
```

I tested using stock lazyvim starter, with enabled extra `util.startuptime`, and my Lazyvim [fork](https://github.com/abeldekat/LazyVim) in `~/projects/plugins/`

```lua
  dev = {
    path = "~/projects/plugins",
    patterns = { "LazyVim" },
  },
```

To all involved: Is it working? Is it still possible to override `vim.opt.clipboard` in the local `options.lua`?

It would be very nice if a windows user could test as well.

**Reply by dpetka2001** [reply]
I just put the changes in the same place in my local forked repo and it works as expected. Can also change `vim.o.clipboard` from user configuration `options.lua`. I tested it with `vim.o.clipboard = ""` and it correctly unset the option, whereas the default option is set to `unnamedplus`.

**Reply by folke** [reply]
Would be better to move the actual loading to line 183.
Right after `M.load("keymaps")`.
I suspect that while `:StartupTime` shows an improvement, even in a `vim.schedule` it might still execute before `UIEnter`? Or maybe not.

**Reply by folke** [reply]
Feel free to make a PR.
I don't see any use-case where the clipboard should be ready right away.

Just make sure to also then check again that `clipboard` is still `""`

**Reply by abeldekat** [reply]
[PR](https://github.com/LazyVim/LazyVim/pull/4120)

**Comment by RayGuo-ergou** [comment]
Thank you so much for sharing this. I added this into my config and I can literally feel the different without any testing. 🥳

**Reply by dpetka2001** [reply]
If you use LazyVim, then you don't need to add this into your config. It's been merged into LazyVim.

**Reply by abeldekat** [reply]
Thanks for the feedback! This is what I was hoping for...

May I ask: Is your clipboard provider `xsel` or `pbcopy`? Do you work on an older laptop, with a HDD perhaps?

Also, you can remove the code from your config, because Lazyvim v12.34.0 has the code built-in.

**Reply by abeldekat** [reply]
Good morning, @dpetka2001....)

**Reply by dpetka2001** [reply]
Morning @abeldekat :D

**Reply by RayGuo-ergou** [reply]
> If you use LazyVim, then you don't need to add this into your config. It's been merged into LazyVim.

> Also, you can remove the code from your config, because Lazyvim v12.34.0 has the code built-in.

Yes I saw the PR, but I am not using LazyVim ( but keep copying code from here of course 😁 )

> May I ask: Is your clipboard provider xsel or pbcopy? Do you work on an older laptop, with a HDD perhaps?

I am using `xsel` and my setup is window wsl2 ( probably extra process to make the clipboard work as the value would pass to window's clipboard eventually ) my PC is fairly good tho.

**Reply by abeldekat** [reply]
It's interesting to know that it's noticeable for `wsl` users on a more recent device. Thanks!

> Yes I saw the PR, but I am not using LazyVim ( but keep copying code from here of course 😁 )

I do the same. `LazyVim` is still a huge inspiration.
